models/buffer_disk.go

```diff
@@ -0,0 +1,151 @@
+package models
+
+import (
+	"bytes"
+	"encoding/gob"
+	"fmt"
+
+	"github.com/influxdata/telegraf"
+	"github.com/tidwall/wal"
+)
+
+type DiskBuffer struct {
+	BufferStats
+
+	walFile *wal.Log
+}
+
+func NewDiskBuffer(name string, capacity int, path string, stats BufferStats) *DiskBuffer {
+	// todo capacity
+	walFile, err := wal.Open(path+"/"+name, nil)
+	if err != nil {
+		return nil // todo error handling
+	}
+	return &DiskBuffer{
+		BufferStats: stats,
+		walFile:     walFile,
+	}
+}
+
+func (b *DiskBuffer) Len() int {
+	return int(b.writeIndex() - b.readIndex())
+}
+
+// readIndex is the first index to start reading metrics from, or the head of the buffer
+func (b *DiskBuffer) readIndex() uint64 {
+	index, err := b.walFile.FirstIndex()
+	if err != nil {
+		panic(err) // can only occur with a corrupt wal file
+	}
+	return index
+}
+
+// writeIndex is the first index to start writing metrics to, or the tail of the buffer
+func (b *DiskBuffer) writeIndex() uint64 {
+	index, err := b.walFile.LastIndex()
+	if err != nil {
+		panic(err) // can only occur with a corrupt wal file
+	}
+	return index
+}
+
+func (b *DiskBuffer) Add(metrics ...telegraf.Metric) int {
+	// one metric to write, can write directly
+	if len(metrics) == 1 {
+		if b.addSingle(metrics[0]) {
+			return 1
+		}
+		return 0
+	}
+
+	// multiple metrics to write, batch them
+	return b.addBatch(metrics)
+}
+
+func (b *DiskBuffer) addSingle(metric telegraf.Metric) bool {
+	err := b.walFile.Write(b.writeIndex(), b.metricToBytes(metric))
+	metric.Accept()
+	if err == nil {
+		b.metricAdded()
+		return true
+	}
+	return false
+}
```

Comment:
Is it really worth the code duplication or should we use the batch-interface instead?